src/core/storage/storage.js

```diff
@@ -32,65 +87,126 @@ export default class GlobalStorage {
    * @param {string} key The storage key to set
    * @param {*} data The data to set
    */
-  set (key, data) {}
+  set (key, data) {
+    if (key === undefined || key === null || typeof key !== 'string') {
+      throw new AnswersStorageError('Invalid storage key provided', key, data);
+    }
+
+    if (typeof data === 'undefined') {
+      throw new AnswersStorageError('Invalid storage key provided', key, data);
+    }
+
+    this.storage.set(key, data);
+    this._callListeners('update', key);
+  }
 
   /**
    * Updates the storage and adds the entry to a persistent storage
-   * buffer. The entry is not yet added to the persistent storage.
+   * buffer. The entry is not added to the persistent storage until
+   * the buffer is flushed.
    *
    * @param {string} key The storage key to set
    * @param {*} data The data to set
    */
-  setWithPersist (key, data, replaceHistory) {}
+  setWithPersist (key, data) {
+    this.set(key, data);
+
+    let serializedData = data;
+    if (typeof data !== 'string') {
+      serializedData = JSON.stringify(data);
+    }
+
+    this.persistentStorageBuffer.set(key, serializedData);
+  }
 
   /**
    * Flushes the persistent storage buffer and adds all entries
    * to the persistent storage.
    */
-  flushPersist () {}
+  flushPersist () {
+    this.persistentStorage = this.persistentStorageBuffer;
+
+    this.persistentStorageBuffer = new DefaultPersistentStorage(this.popListener);
+    this.persistentStorage.getAll().forEach((value, key) => {
+      this.persistentStorageBuffer.set(key, value);
+    });
```

Comment:
When I think of flushing a buffer, I assume that the buffer is emptied. However that's not the case here, is it? Since the persistentStorageBuffer is immediately updated with all the values from the new persistentStorage?

Reply:
Persistent storage buffer is under construction, updates pending. I agree that it feels like flush should empty the buffer.